File: composite/java/CompoundOrder.java
import java.util.Vector;

abstract class Order {

	private double price;

	protected Order(double price) {
		this.price = price;
	}

	protected double getPrice() {
		return this.price;
	}
}

final class Hammer extends Order {

	Hammer(double price) {
		super(price);
	}
}

final class Phone extends Order {

	Phone(double price) {
		super(price);
	}
}

final class Earphone extends Order {

	Earphone(double price) {
		super(price);
	}
}

final class Charger extends Order {

	Charger(double price) {
		super(price);
	}
}

final class Receip extends Order {

	Receip(double price) {
		super(price);
	}
}

final class Printer extends Order {

	Printer(double price) {
		super(price);
	}
}

final class Box extends Order {

	private Vector<Order> orders;

	Box() {
		super(0.0);
		this.orders = new Vector();
	}

	public void add(Order order) {
		this.orders.add(order);
	}

	@Override
	public double getPrice() {
		return orders.stream().mapToDouble(Order::getPrice).sum();
	}
}

final class CompoundOrder {


	public static void main(String[] args) {
		Order hammer = new Hammer(130.50);
		Order phone = new Phone(2500.0);
		Order earphone = new Earphone(50.0);
		Order charger = new Charger(70.0);
		Order receip = new Receip(30.14);
		Order printer = new Printer(900.0);

		Box container = new Box();
		Box phoneBox = new Box();
		phoneBox.add(phone);
		phoneBox.add(earphone);
		phoneBox.add(charger);

		Box hammerBox = new Box();
		hammerBox.add(hammer);

		container.add(phoneBox);
		container.add(hammerBox);
		container.add(receip);
		container.add(printer);

		System.out.println("Total order price: "+container.getPrice());
	}
}

File: composite/kotlin/compound_order.kt
interface Order {

	val price: Double
}

class Hammer(override val price: Double) : Order

class Phone(override val price: Double) : Order

class Earphone(override val price: Double) : Order

class Charger(override val price: Double) : Order

class Receip(override val price: Double) : Order

class Printer(override val price: Double) : Order

class Box : Order {

	val orders: MutableList<Order>
	override val price: Double
		get() = this@Box.orders.sumOf { it.price }

	init {
		this.orders = mutableListOf()
	}

	fun add(order: Order): Unit {
		this.orders.add(order)
	}
}

fun main(): Unit {
	val hammer: Order = Hammer(130.50)
	val phone: Order = Phone(2500.0)
	val earphone: Order = Earphone(50.0)
	val charger: Order = Charger(70.0)
	val receip: Order = Receip(30.14)
	val printer: Order = Printer(900.0)

	val container: Box = Box()
	val phoneBox: Box = Box()
	phoneBox.add(phone)
	phoneBox.add(earphone)
	phoneBox.add(charger)

	val hammerBox: Box = Box()
	hammerBox.add(hammer)

	container.add(phoneBox)
	container.add(hammerBox)
	container.add(receip)
	container.add(printer)

	println("Total order price: ${container.price}")
}